File: packages/mobile_android/app/src/main/java/dev/verdeai/app/ViewCache.kt
package dev.verdeai.app

import dev.verdeai.core.*
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString

/** Last live core projections for one host, shown only until that host's core has synced. */
@Serializable
internal data class CachedViews(val version: Int = 1, val saved_at_ms: Long, val home: HomeView, val workspaces: WorkspacesView)

/**
 * Warm-start cache in its own encrypted, no-backup store. Content (titles, paths) is never
 * logged. Sign-out and removal clear it; unreadable or oversized entries are discarded.
 */
internal class ViewCache(private val store: SecureStore) {
    private val lock = Mutex()
    suspend fun load(hostId: String): CachedViews? = try {
        store.get(key(hostId))?.let { CoreJson.decodeFromString<CachedViews>(it) }?.takeIf { it.version == 1 }
    } catch (e: CancellationException) { throw e } catch (_: Exception) { null }

    /**
     * Returns false when [allowed] no longer holds, the bounded copy is still too large, or the
     * store rejects it. Writes and [clear] are serialized, so a save that races sign-out either
     * observes the wipe in [allowed] or lands before the clear that removes it.
     */
    suspend fun save(hostId: String, views: CachedViews, allowed: () -> Boolean = { true }): Boolean = try {
        val encoded = CoreJson.encodeToString(bounded(views))
        if (encoded.length > MAX_BYTES) false
        else lock.withLock { if (!allowed()) false else { store.put(key(hostId), encoded); true } }
    } catch (e: CancellationException) { throw e } catch (_: Exception) { false }

    suspend fun clear(hostId: String) {
        try { lock.withLock { store.delete(key(hostId)) } } catch (e: CancellationException) { throw e } catch (_: Exception) { }
    }

    private fun bounded(views: CachedViews) = views.copy(
        home = views.home.copy(items = views.home.items.take(MAX_ITEMS)),
        workspaces = views.workspaces.copy(
            items = views.workspaces.items.take(MAX_ITEMS).map { it.copy(panes = it.panes.take(MAX_ITEMS), threads = it.threads.take(MAX_ITEMS)) },
            history = views.workspaces.history.copy(items = views.workspaces.history.items.take(MAX_ITEMS), next_cursor = null)))

    companion object {
        const val MAX_BYTES = 512 * 1024
        const val MAX_ITEMS = 200
        fun key(hostId: String) = "android/1/views/$hostId"
    }
}

File: packages/mobile_ios/App/ViewCache.swift
import Foundation
import CryptoKit

/// Last live core projections for one host, shown only until that host's core has synced.
struct CachedViews: Codable {
    var version: Int = 1
    var saved_at_ms: Int64
    var home: HomeView
    var workspaces: WorkspacesView
}

/// Warm-start cache: one AES-GCM file per host under Application Support,
/// excluded from backup and written with file protection. The key lives only in
/// this device's Keychain (`AfterFirstUnlockThisDeviceOnly`), so a restored or
/// copied file is unreadable. Content (titles, paths) is never logged. Sign-out
/// and removal clear it; unreadable or oversized entries are discarded.
///
/// Main-actor only: saves and clears are serialized with the host state that
/// decides whether a save is still allowed, so a save cannot land after a wipe.
@MainActor
final class ViewCache {
    static let maxBytes = 512 * 1024
    static let maxItems = 200
    static let keyName = "ios/1/view-cache-key"
    let directory: URL
    private let keys: SecureStorage

    init(directory: URL, keys: SecureStorage) {
        self.directory = directory
        self.keys = keys
    }

    static func live() -> ViewCache? {
        guard let support = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first else { return nil }
        return ViewCache(directory: support.appendingPathComponent("ViewCache", isDirectory: true),
                         keys: KeychainStorage(service: "dev.verdeai.app.cache"))
    }

    func load(_ hostID: String) -> CachedViews? {
        guard let url = file(hostID), let sealed = try? Data(contentsOf: url) else { return nil }
        do {
            guard let raw = try keys.get(Self.keyName) else { throw CocoaError(.fileReadCorruptFile) }
            let box = try AES.GCM.SealedBox(combined: sealed)
            let plain = try AES.GCM.open(box, using: SymmetricKey(data: raw), authenticating: Self.aad(hostID))
            let views = try JSONDecoder().decode(CachedViews.self, from: plain)
            guard views.version == 1 else { throw CocoaError(.fileReadCorruptFile) }
            return views
        } catch {
            try? FileManager.default.removeItem(at: url)
            return nil
        }
    }

    /// Returns false when the bounded copy is still too large or storage rejects it.
    @discardableResult
    func save(_ hostID: String, _ views: CachedViews) -> Bool {
        guard let url = file(hostID) else { return false }
        do {
            let plain = try JSONEncoder().encode(Self.bounded(views))
            guard plain.count <= Self.maxBytes else { return false }
            let box = try AES.GCM.seal(plain, using: key(), authenticating: Self.aad(hostID))
            guard let sealed = box.combined else { return false }
            try prepareDirectory()
            try sealed.write(to: url, options: [.atomic, .completeFileProtectionUntilFirstUserAuthentication])
            var file = url
            var values = URLResourceValues()
            values.isExcludedFromBackup = true
            try file.setResourceValues(values)
            return true
        } catch { return false }
    }

    func clear(_ hostID: String) {
        guard let url = file(hostID) else { return }
        try? FileManager.default.removeItem(at: url)
    }

    static func bounded(_ views: CachedViews) -> CachedViews {
        var copy = views
        copy.home.items = Array(views.home.items.prefix(maxItems))
        copy.workspaces.items = views.workspaces.items.prefix(maxItems).map { workspace in
            var bounded = workspace
            bounded.panes = Array(workspace.panes.prefix(maxItems))
            bounded.threads = Array(workspace.threads.prefix(maxItems))
            return bounded
        }
        copy.workspaces.history.items = Array(views.workspaces.history.items.prefix(maxItems))
        copy.workspaces.history.next_cursor = nil
        return copy
    }

    private static func aad(_ hostID: String) -> Data { Data("verde-view-cache/1/\(hostID)".utf8) }

    private func file(_ hostID: String) -> URL? {
        guard HostsModel.validID(hostID) else { return nil }
        return directory.appendingPathComponent("\(hostID).views", isDirectory: false)
    }

    private func key() throws -> SymmetricKey {
        if let raw = try keys.get(Self.keyName), raw.count == 32 { return SymmetricKey(data: raw) }
        let key = SymmetricKey(size: .bits256)
        try keys.put(Self.keyName, value: key.withUnsafeBytes { Data($0) })
        return key
    }

    private func prepareDirectory() throws {
        var url = directory
        guard !FileManager.default.fileExists(atPath: url.path) else { return }
        try FileManager.default.createDirectory(at: url, withIntermediateDirectories: true,
            attributes: [.protectionKey: FileProtectionType.completeUntilFirstUserAuthentication])
        var values = URLResourceValues()
        values.isExcludedFromBackup = true
        try url.setResourceValues(values)
    }
}
